### ISSUE TITLE
"Unable to unroll loop" error compiling generated HLSL code that writes to a matrix row in a loop

### ISSUE THREAD

**Issue by pixelspark** [gfx-rs/naga#2350]
We are using `wgpu` in `wonnx` to run machine learning models using compute shaders - as I understand it the WGSL code is translated to HLSL (and MSL, etc) through `naga`. All of this has been working great so far, your work is much appreciated!

While testing code that works fine on macOS I nevertheless encountered an issue on Windows (https://github.com/webonnx/wonnx/issues/166). The error is the following:

````
In Device::create_compute_pipeline
Internal error: FXC D3DCompile error (0x80004005): (81,21-35): warning X3550: array reference cannot be used as an l-value; not natively addressable, forcing loop to unroll
(35,5-15): error X3511: unable to unroll loop, loop does not appear to terminate in a timely manner (329 iterations) or unrolled 
````

The linked issue contains both the input WGSL code as well as the generated HLSL code. It appears the issue is the following:

````hlsl
// tmpsum and product are both matNxN where N = kernel_size
for(var k: u32 = 0u; k < {{ k_chunks }}u; k = k + 1u) {
  // ...
 for(var index_mat: u32 = 0u; index_mat < {{ kernel_size }}u; index_mat = index_mat + 1u) {
    tmpsum[index_mat] = tmpsum[index_mat] + product[index_mat];
 }
}
````

I managed to resolve it (see https://github.com/webonnx/wonnx/commit/c772653fe97c0e9b523cdf530a71bff66ec51176) by replacing the inner loop with simply `product = mat_right * mat_left;`. This is better regardless of this issue. Nevertheless you may be interested to know about this error as I suspect it may occur in different situations as well.

(Note that I am now seeing some other error, see the linked issue. If anyone here happens to have an idea about that one, I'd be very interested to hear)

**Comment by teoxoy** [comment]
> Apparently HLSL requires the inner loop to be unrolled as it apparently disallows indexing into `matNxN` dynamically (while WGSL allows this). It helpfully applies a workaround (unroll the loop) but inadvertently seems to unroll *both* the inner and outer loop.

Well, this is unfortunate... We might have to come up with a workaround for this (unroll the inner loop ourselves?) but I don't think it's high priority right now.

Possibly related https://github.com/gfx-rs/naga/issues/1702.

**Comment by pixelspark** [comment]
Yeah, not sure about what the right solution would be in this case. Fundamentally WGSL is allowing something that HLSL does not support (dynamic indexing of a matrix row). Notifying the user by throwing an error in this case seems appropriate to me. Working around the issue entirely may not be possible nor worthwhile. 

Perhaps the HLSL compiler can be tricked into unrolling the inner loop but not the outer one by application of attributes (see HLSL documentation - 'unroll' is one of the attributes). That would fix this specific case but dynamic indexing in matrixes (and perhaps vectors) could also occur in other usage scenarios.
